modules/calib3d/src/ap3p.cpp

```diff
@@ -0,0 +1,369 @@
+}
+
+inline double ap3p::vect_norm(const double *a) {
+  return sqrt(a[0] * a[0] + a[1] * a[1] + a[2] * a[2]);
+}
+
+inline void ap3p::vect_scale(const double s, const double *a, double *result) {
+  result[0] = a[0] * s;
+  result[1] = a[1] * s;
+  result[2] = a[2] * s;
+}
+
+inline void ap3p::vect_sub(const double *a, const double *b, double *result) {
+  result[0] = a[0] - b[0];
+  result[1] = a[1] - b[1];
+  result[2] = a[2] - b[2];
+}
+
+inline void ap3p::vect_divide(const double *a, const double d, double *result) {
+  result[0] = a[0] / d;
+  result[1] = a[1] / d;
+  result[2] = a[2] / d;
+}
+
+inline void ap3p::mat_mult(const double a[3][3], const double b[3][3], double result[3][3]) {
+  result[0][0] = a[0][0] * b[0][0] + a[0][1] * b[1][0] + a[0][2] * b[2][0];
+  result[0][1] = a[0][0] * b[0][1] + a[0][1] * b[1][1] + a[0][2] * b[2][1];
+  result[0][2] = a[0][0] * b[0][2] + a[0][1] * b[1][2] + a[0][2] * b[2][2];
+
+  result[1][0] = a[1][0] * b[0][0] + a[1][1] * b[1][0] + a[1][2] * b[2][0];
+  result[1][1] = a[1][0] * b[0][1] + a[1][1] * b[1][1] + a[1][2] * b[2][1];
+  result[1][2] = a[1][0] * b[0][2] + a[1][1] * b[1][2] + a[1][2] * b[2][2];
+
+  result[2][0] = a[2][0] * b[0][0] + a[2][1] * b[1][0] + a[2][2] * b[2][0];
+  result[2][1] = a[2][0] * b[0][1] + a[2][1] * b[1][1] + a[2][2] * b[2][1];
+  result[2][2] = a[2][0] * b[0][2] + a[2][1] * b[1][2] + a[2][2] * b[2][2];
+}
+
+bool ap3p::solve(cv::Mat &R, cv::Mat &tvec, const cv::Mat &opoints, const cv::Mat &ipoints) {
+  CV_INSTRUMENT_REGION()
+
+  double rotation_matrix[3][3], translation[3];
+  std::vector<double> points;
+  if (opoints.depth() == ipoints.depth()) {
+    if (opoints.depth() == CV_32F)
+      extract_points<cv::Point3f, cv::Point2f>(opoints, ipoints, points);
+    else
+      extract_points<cv::Point3d, cv::Point2d>(opoints, ipoints, points);
+  } else if (opoints.depth() == CV_32F)
+    extract_points<cv::Point3f, cv::Point2d>(opoints, ipoints, points);
+  else
+    extract_points<cv::Point3d, cv::Point2f>(opoints, ipoints, points);
+
+  bool result = solve(rotation_matrix, translation, points[0], points[1], points[2], points[3], points[4], points[5],
+                      points[6], points[7], points[8], points[9], points[10], points[11], points[12], points[13],
+                      points[14],
+                      points[15], points[16], points[17], points[18], points[19]);
+  cv::Mat(3, 1, CV_64F, translation).copyTo(tvec);
+  cv::Mat(3, 3, CV_64F, rotation_matrix).copyTo(R);
+  return result;
+}
+
+bool
+ap3p::solve(double (*R)[3], double *t, double mu0, double mv0, double X0, double Y0, double Z0, double mu1, double mv1,
+            double X1, double Y1, double Z1, double mu2, double mv2, double X2, double Y2, double Z2, double mu3,
+            double mv3, double X3, double Y3, double Z3) {
+  double Rs[4][3][3], ts[4][3];
+
+  int n = solve(Rs, ts, mu0, mv0, X0, Y0, Z0, mu1, mv1, X1, Y1, Z1, mu2, mv2, X2, Y2, Z2);
```

Comment:
I need to check if it is possible to determine the number of solutions easily. 
The current code always returns 4 because it treats complex roots as real ones. This is based on the concern that sometimes a valid solution may correspond to a complex root with small imaginary part due to noise.